lag metric reporting incorrect statistics

Sometimes, some configurations will report a high lag time, which triggers the `lag` state in `sr3 status`.

Example

```
Component/Config                              Processes   Connection        Lag                              Rates
                                              State   Run Retry  msg data   Que   LagMax   LagAvg  Last  %rej     pubsub   messages     RxData     TxData
sarra/get-cmems-oceanobs                      lag     1/1     0 100%   0%     0  171.24s   90.80s 2.43s  0.0%    564 B/s      0 m/s 40.3 KiB/s      0 B/s
sarra/get_airnow_http                         lag     1/1     1 100%  67%     0 541312.32s 541312.32s 4196s  0.0%      0 B/s      0 m/s      0 B/s      0 B/s
sarra/get_airnow_prov_depot                   lag     2/2    13 100%  69%     0 586328.54s 310211.80s 6940s  0.0%      0 B/s      0 m/s      0 B/s      0 B/s
sarra/get_marine_weather                      lag     1/1    53 100%  72%     0 533804.73s 212781.40s 8549s  0.0%      0 B/s      0 m/s      0 B/s      0 B/s
sarra/get_rvas_rem_pxpaz                      lag     1/1     1 100%  91%     0 270340.14s 270340.14s 6424s  0.0%      0 B/s      0 m/s      0 B/s      0 B/s
```

The first value looks to make sense. The other 4 values span from 3 to 6 days which don't make sense when looking at their `last` values.
So far this problem looks to only be affecting sarra components. Still unsure why

We checked the output of `sr3 dump` it also looked to report a similar story. `lagMax` and `lagTotal` are of huge values

```
"get_marine_weather" :  {"instance_pids": {"1": 16077}, "queueName": "q_WXO-DD.sarra.get_marine_weather.ddsr.cmc.ec.gc.ca", "has_state": false, "noVip": null, "instance_metrics": {"1": {"flow": {"stop_requested": false, "last_housekeeping": 1725462788.3809488, "transferConnected": true, "transferConnectStart": 1725462900.03323, "transferConnectTime": 62.7959840297699, "transferRxBytes": 0, "transferTxBytes": 0, "transferRxFiles": 0, "transferTxFiles": 0, "last_housekeeping_cpuTime": 27725.94, "cpuTime": 12.299999999999272, "next_housekeeping": 1725463088.3809488, "msgRate": 0.005914938262384488, "msgRateCpu": 0.1625919536811215, "current_sleep": 6.4}, "post.message": {"connected": true, "disconnectLast": 0, "disconnectTime": 0, "disconnectCount": 0, "rxByteCount": 0, "rxGoodCount": 0, "rxBadCount": 0, "txByteCount": 0, "txGoodCount": 0, "txBadCount": 0, "txLast": "20240904T124839.0826826096"}, "gather.message": {"connected": true, "disconnectLast": 0, "disconnectTime": 0, "disconnectCount": 0, "rxByteCount": 0, "rxGoodCount": 0, "rxBadCount": 0, "txByteCount": 0, "txGoodCount": 0, "txBadCount": 0, "brokerQueuedMessageCount": 0, "rxLast": "20240904T124838.50991869"}, "retry": {"msgs_in_download_retry": 53, "msgs_in_post_retry": 0}, 
"log": {"lagMax": 534114.6203896999, "lagTotal": 11293876.948493004, "lagMessageCount": 53, "rejectCount": 0}, "sarracenia.transfer": {"byteRateInstant": 2300926.898529939}, "status": {"mtime": 1725462900.121718}}}, "logAge": {"1": 65.95925784111023}, "metrics": {"byteRate": 0.0, "cpuTime": 12.299999999999272, "rejectCount": 0, "last_housekeeping": 1725462788.3809488, "messagesQueued": 0, "lagMean": 213092.01789609442, "latestTransfer": "8786s", "rejectPercent": 0, "transferRxByteRate": 0.0, "transferTxByteRate": 0.0, "rxByteCount": 0, "rxGoodCount": 0, "rxBadCount": 0, "txByteCount": 0, "txGoodCount": 0, "txBadCount": 0, "lagMax": 534114.6203896999, "lagTotal": 11293876.948493004, "lagMessageCount": 53, "disconnectTime": 0.0, "transferConnectTime": 62.7959840297699, "transferRxLast": 0, "transferTxLast": 0, "rxLast": 1725454119.0826826, "txLast": 1725454119.0826826, "transferRxBytes": 0, "transferRxFiles": 0, "transferTxBytes": 0, "transferTxFiles": 0, "msgs_in_post_retry": 0, "msgs_in_download_retry": 53, "brokerQueuedMessageCount": 0, "time_base": 116.35145330429077, "byteTotal": 0, "msgRate": 0.0, "msgRateCpu": 0.0, "retry": 53, "transferLast": 0, "connectPercent": 100, "byteConnectPercent": 53, "messageLast": 1725454119.0826826, "transferRxFileRate": 0.0, "transferTxFileRate": 0.0}, "missing_instances": [], "hung_instances": [], "resource_usage": {"uss": 60354560, "rss": 76558336, "vms": 325898240, "user_cpu": 27219.12, "system_cpu": 519.13}}

```


## Work around
- ~~The work around for this is to increase the values of [runStateThreshold_lag](https://metpx.github.io/sarracenia/Reference/sr3_options.7.html#runstatethreshold-lag-interval-default-30) to an astronomical value on each of the configurations so that the status is never reported.~~
- do a cleanup.
- make the retry_ttl lower ? so it will age out faster?
- retry off (prevent retries completely) if this is a poll and the data is old, then maybe not worth retrieving.

so... I took a look at sarra/get_airnow_http:
```
2024-09-04 00:02:34,498 [DEBUG] sarracenia.diskqueue msg_get_from_file DEBUG /local/home/sarra/.cache/sr3/sarra/get_airnow_http/diskqueue_work_retry_01 open r
ead
2024-09-04 00:02:34,500 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 473281.70 ) https://s3-us-west-1.amazonaws.com//files.airnowtech.org/airnow/
```

It is retrieving a message from the retry queue, and then the retrieve fails, and it puts it back on the retry queue, every five minutes.

```
2024-09-04 00:02:34,504 [DEBUG] sarracenia.transfer.https __open__ https://s3-us-west-1.amazonaws.com///files.airnowtech.org/airnow/today/HourlyData_202408291
1.dat
2024-09-04 00:02:34,772 [ERROR] sarracenia.transfer.https __open__ Download failed 4 https://s3-us-west-1.amazonaws.com//files.airnowtech.org/airnow/today/Hou
rlyData_2024082911.dat
2024-09-04 00:02:34,772 [ERROR] sarracenia.transfer.https __open__ Server couldn't fulfill the request. Error code: 404, Not Found

```
so that is causing the huge lag report.

Question:  should 404 cause us to give up on retries?   If it doesn't have it now, when would we expect it to have it.  the new "stat" code just merged got rid of some useless cases when putting stuff on a retry queue is useless... this might be another.    otoh, the danger is that, when a server is sick, it gives bogus 404's that eventually get fixed.

Probably the most flexible way is to have a set of codes that make us retry, and others that make us give up... so it's analyst tunable.

I found a similar lag output in the `sr3 status`, this time in a sender. So it's getting the files locally.

```
ddsr-cmc-stage02:                                    State   Run Retry  msg data   Que   LagMax   LagAvg  Last  %rej     pubsub   messages     RxData     TxData
ddsr-cmc-stage02: sender/dnd_to_ice                  lag     1/1    17 100%   0%     0 563059.95s 212969.82s >9999  0.0%      0 B/s      0 m/s      0 B/s      0 B/s
```

It's reporting the lag being over 6 days again. But the last sent file was about 14 hours ago.
There's also old files getting retried from the retry queue, despite them not existing anymore locally.

```
ddsr-cmc-stage01: 2024-09-06 18:50:40,445 [ERROR] sarracenia.flow send could not chdir to /apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05 to write: [Errno 2] No such file or directory: '/apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05'
ddsr-cmc-stage01: 2024-09-06 18:50:40,445 [WARNING] sarracenia.flow do_send sending again, attempt 2
ddsr-cmc-stage01: 2024-09-06 18:50:40,445 [ERROR] sarracenia.flow send could not chdir to /apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05 to write: [Errno 2] No such file or directory: '/apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05'
ddsr-cmc-stage01: 2024-09-06 18:50:40,445 [WARNING] sarracenia.flow do_send sending again, attempt 3
ddsr-cmc-stage01: 2024-09-06 18:50:40,445 [ERROR] sarracenia.flow send could not chdir to /apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05 to write: [Errno 2] No such file or directory: '/apps/sarra/public_data/20240903/DND/ICE/sfc-prog/05'

```

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -2294,7 +2294,15 @@ class Flow:
         return 1
 
     # generalized send...
-    def send(self, msg, options):
+    def send(self, msg, options) -> int:
+        """
+           send/transfer one file based on message, return int:
+
+           return 0 -- failed temporarily, retry later.
+           return >0 -- OK download successful.
+           return <0 -- download failed permanently, retry not useful.
+        """
+
         self.o = options
         sendTo=self.o.sendTo 
         logger.debug( f"{self.scheme}_transport sendTo: {sendTo}" )
@@ -2306,14 +2314,16 @@ class Flow:
             for plugin in self.plugins['send']:
                 try:
                     ok = plugin(msg)
-                    if type(ok) is not bool:
-                        logger.error( f"{plugin} returned {type(ok)}. Should return boolean" )
+                    if type(ok) not in [ bool, int ]:
+                        logger.error( f"{plugin} returned {type(ok)}. Must return boolean or integer, discarding." )
+                        return -1
                 except Exception as ex:
                     logger.error( f'flowCallback plugin {plugin} crashed: {ex}' )
                     logger.debug( "details:", exc_info=True )
 
-                if not ok: return False
-            return True
+                if type(ok) is bool and not ok: return 0
+                elif type(ok) is int: return ok
+            return 1
 
         if self.o.baseDir:
             local_path = self.o.variableExpansion(self.o.baseDir,
@@ -2342,9 +2352,8 @@ class Flow:
             try:
                 os.chdir(local_dir)
             except Exception as ex:
-                logger.error("could not chdir to %s to write: %s" % (local_dir, ex))
-                return False
-
+                logger.error("could not chdir locally to %s: %s" % (local_dir, ex))
+                return -1
         try:
 
             if not self.o.dry_run:
@@ -2361,7 +2370,7 @@ class Flow:
                     self.proto[self.scheme] = sarracenia.transfer.Transfer.factory( self.scheme, options)
     
                     ok = self.proto[self.scheme].connect()
-                    if not ok: return False
+                    if not ok: return 0
                     self.cdir = None
                     self.metrics['flow']['transferConnected'] = True
                     self.metrics['flow']['transferConnectStart'] = time.time() 
@@ -2382,7 +2391,7 @@ class Flow:
                                msg['blocks']['method'] == 'inplace'):
                 logger.error("%s, inplace part file not supported" %
                              self.scheme)
-                return False
+                return -1
 
             #=================================
             # if umask, check that the protocol supports it ...
@@ -2414,7 +2423,7 @@ class Flow:
                         cwd = self.proto[self.scheme].getcwd()
                     except Exception as ex:
                         logger.error( f"could not getcwd on {sendTo} : {ex}" )
-                        return False
+                        return 0
 
             if cwd != new_dir:
                 logger.debug("%s_transport send cd to %s" %
@@ -2424,7 +2433,7 @@ class Flow:
                         self.proto[self.scheme].cd_forced(new_dir)
                     except Exception as ex:
                         logger.error( f"could not chdir to {sendTo} {new_dir}: {ex}" )
-                        return False
+                        return -1
 
             #=================================
             # delete event
@@ -2446,14 +2455,14 @@ class Flow:
                                     self.proto[self.scheme].delete(new_file)
                                 except Exception as ex:
                                     logger.error( f"could not delete {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                                    return False
+                                    return -1
 
                         msg.setReport(201, f'file or directory removed')
                         self.metrics['flow']['transferTxFiles'] += 1
                         self.metrics['flow']['transferTxLast'] = msg['report']['timeCompleted']
-                        return True
+                        return 1
                     logger.error("%s, delete not supported" % self.scheme)
-                    return False
+                    return -1
 
                 if 'rename' in msg['fileOp'] :
                     if hasattr(self.proto[self.scheme], 'delete'):
@@ -2463,14 +2472,14 @@ class Flow:
                                 self.proto[self.scheme].rename(msg['fileOp']['rename'], new_file)
                             except Exception as ex:
                                 logger.error( f"could not rename {sendTo} (in {msg['new_dir']} {msg['fileOp']['rename']} to {new_file}: {ex}" )
-                                return False
+                                return -1
 
                         msg.setReport(201, f'file renamed')
                         self.metrics['flow']['transferTxFiles'] += 1
                         self.metrics['flow']['transferTxLast'] = msg['report']['timeCompleted']
-                        return True
+                        return 1
                     logger.error("%s, delete not supported" % self.scheme)
-                    return False
+                    return -1
 
                 if 'directory' in msg['fileOp'] :
                     if 'contentType' not in msg:
@@ -2482,13 +2491,13 @@ class Flow:
                                 self.proto[self.scheme].mkdir(new_file)
                             except Exception as ex:
                                 logger.error( f"could not mkdir {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                                return False
+                                return 0
                         msg.setReport(201, f'directory created')
                         self.metrics['flow']['transferTxFiles'] += 1
                         self.metrics['flow']['transferTxLast'] = msg['report']['timeCompleted']
-                        return True
+                        return 1
                     logger.error("%s, mkdir not supported" % self.scheme)
-                    return False
+                    return -1
 
 
                 #=================================
@@ -2505,10 +2514,10 @@ class Flow:
                                 self.proto[self.scheme].link(msg['fileOp']['hlink'], new_file)
                             except Exception as ex:
                                 logger.error( f"could not link {sendTo} in {msg['new_dir']}{os.sep}{msg['fileOp']['hlink']} to {new_file}: {ex}" )
-                                return False
-                        return True
+                                return 0
+                        return 1
                     logger.error("%s, hardlinks not supported" % self.scheme)
-                    return False
+                    return -1
                 elif 'link' in msg['fileOp']:
                     if 'contentType' not in msg:
                         msg['contentType'] = 'text/link'
@@ -2519,13 +2528,13 @@ class Flow:
                                 self.proto[self.scheme].symlink(msg['fileOp']['link'], new_file)
                             except Exception as ex:
                                 logger.error( f"could not symlink {sendTo} in {msg['new_dir']} {msg['fileOp']['link']} to {new_file}: {ex}" )
-                                return False
+                                return 0
                         msg.setReport(201, f'file linked')
                         self.metrics['flow']['transferTxFiles'] += 1
                         self.metrics['flow']['transferTxLast'] = msg['report']['timeCompleted']
-                        return True
+                        return 1
                     logger.error("%s, symlink not supported" % self.scheme)
-                    return False
+                    return -1
 
             #=================================
             # send event
@@ -2537,7 +2546,7 @@ class Flow:
                     "product collision or base_dir not set, file %s does not exist"
                     % local_file)
                 time.sleep(0.01)
-                return False
+                return 0
             elif 'size' not in msg:
                 msg['size'] = os.path.getsize(local_file)
 
@@ -2580,7 +2589,7 @@ class Flow:
                             len_written = self.proto[self.scheme].put( msg, local_file, new_file)
                 except Exception as ex:
                     logger.error( f"could not send {local_dir}{os.sep}{local_file} to inflight=None {sendTo} {msg['new_dir']} ... {new_file}: {ex}" )
-                    return False
+                    return 0
                 
             elif (('blocks' in msg)
                   and (msg['blocks']['method'] == 'inplace')):
@@ -2589,8 +2598,8 @@ class Flow:
                         self.proto[self.scheme].put(msg, local_file, new_file, offset,
                                             new_offset, msg['size'])
                     except Exception as ex:
-                        logger.error( f"could not send {local_dir}{os.sep}{local_file} inplace {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        logger.error( f"could not send {local_dir}{os.sep}{local_file} inplace {sendTo} {msg['new_dir']} ... {new_file}: {ex}" )
+                        return 0
 
             elif inflight == '.':
                 new_inflight_path = '.' + new_file
@@ -2604,12 +2613,12 @@ class Flow:
                                 msg, local_file, new_inflight_path)
                     except Exception as ex:
                         logger.error( f"could not send {local_dir}{os.sep}{local_file} inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                     try:
                         self.proto[self.scheme].rename(new_inflight_path, new_file)
                     except Exception as ex:
                         logger.error( f"could not rename inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                 else:
                     len_written = msg['size']
 
@@ -2624,12 +2633,12 @@ class Flow:
                             len_written = self.proto[self.scheme].put(msg, local_file, new_inflight_path)
                     except Exception as ex:
                         logger.error( f"could not send {local_dir}{os.sep}{local_file} inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                     try:
                         self.proto[self.scheme].rename(new_inflight_path, new_file)
                     except Exception as ex:
                         logger.error( f"could not rename inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
             elif options.inflight[-1] == '/':
                 if not self.o.dry_run:
                     try:
@@ -2649,12 +2658,12 @@ class Flow:
                                 msg, local_file, new_inflight_path)
                     except Exception as ex:
                         logger.error( f"could not send {local_dir}{os.sep}{local_file} inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                     try:
                         self.proto[self.scheme].rename(new_inflight_path, new_file)
                     except Exception as ex:
                         logger.error( f"could not rename inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                 else:
                     len_written = msg['size']
             elif inflight == 'umask':
@@ -2669,12 +2678,12 @@ class Flow:
                                 msg, local_file, new_file)
                     except Exception as ex:
                         logger.error( f"could not send {local_dir}{os.sep}{local_file} inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                     try:
                         self.proto[self.scheme].put(msg, local_file, new_file)
                     except Exception as ex:
                         logger.error( f"could not rename inflight={inflight} {sendTo} {msg['new_dir']}/{new_file}: {ex}" )
-                        return False
+                        return 0
                 else:
                     len_written = msg['size']
 
@@ -2693,7 +2702,7 @@ class Flow:
                         (local_path, str_range, new_dir, new_file, offset,
                          offset + msg['size'] - 1))
 
-            return True
+            return 1
 
         except Exception as err:
 
@@ -2724,7 +2733,7 @@ class Flow:
                          msg['new_file'])
             logger.debug('Exception details: ', exc_info=True)
 
-            return False
+            return 0
 
     # set_local_file_attributes
     def set_local_file_attributes(self, local_file, msg):
@@ -2867,13 +2876,16 @@ class Flow:
                 if i != 1:
                     logger.warning("sending again, attempt %d" % i)
 
-                ok = self.send(msg, self.o)
-                if ok:
+                retval = self.send(msg, self.o)
+                if retval > 0:
                     self.worklist.ok.append(msg)
+                    break
+                elif retval < 0:
+                    self.worklist.rejected.append(msg)
                     break
 
                 i = i + 1
-            if not ok:
+            if retval == 0:
                 self.worklist.failed.append(msg)
         self.worklist.incoming = []
 

--- a/sarracenia/flowcb/__init__.py
+++ b/sarracenia/flowcb/__init__.py
@@ -180,11 +180,11 @@ class FlowCB:
          to indicate failure to process a message, append to worklist.failed.
          worklist.failed processing should occur in here as it will be zeroed out after this step.
 
-    def send(self,msg) -> bool::
+    def send(self,msg) -> int::
 
          Task: looking at msg['new_dir'], msg['new_file'], and the self.o options perform a transfer
                of a single file.
-               return True on a successful transfer, False otherwise.
+               return >0 on a successful transfer, 0 if failed but temporarily, <0 if failure is permanent.
 
                if self.o.dry_run is set, simulate the output of a send without
                performing it.

